dali/pipeline/executor/executor.cc

```diff
@@ -398,6 +399,38 @@ void Executor<WorkspacePolicy, QueuePolicy>::RunHelper(OpNode &op_node, Workspac
                              ws.GetRequestedBatchSize(i), " <= ", max_batch_size_));
   }
 
+  // If we are using conditional statements try to keep smaller memory footprint by not keeping
+  // the largest batch produced in conditional block as reserved memory.
+  // If the conditionals were detected in this graph (split operator with `_if_stmt` argument)
+  // see if the operator is executed in conditional scope - seeing a partial batch processed
+  // would indicate it in most cases (unless someone mixes partial batches and conditionals).
+  // In such case reset the outputs before adjusting their size so we free the reserved memory
+  // and allocated only batch of currently needed size.
+  if (has_conditionals_) {
+    bool any_tensor_partial = false;
+    if (ws.NumInput() > 0 || HasTensorArgInputs(ws)) {
+      for (int i = 0; i < ws.NumInput(); i++) {
+        any_tensor_partial = any_tensor_partial || ws.GetInputBatchSize(i) < max_batch_size_;
+      }
+      const ArgumentWorkspace &argument_ws = ws;
+      for (const auto &[name, arg] : argument_ws) {
+        any_tensor_partial = any_tensor_partial || arg.tvec->num_samples() < max_batch_size_;
+      }
+    }
+    for (int i = 0; i < spec.NumOutput(); i++) {
+      any_tensor_partial = any_tensor_partial || ws.GetRequestedBatchSize(i) < max_batch_size_;
+    }
+    if (any_tensor_partial) {
+      for (int i = 0; i < spec.NumOutput(); i++) {
+        if (ws.template OutputIsType<CPUBackend>(i)) {
+          ws.template Output<CPUBackend>(i).Reset();
+        } else {
+          ws.template Output<GPUBackend>(i).Reset();
+        }
+      }
+    }
+  }
+
```

Comment:
Maybe this whole block would be more readable as a separate method?

Reply:
done